src/types/merkle_tree_field.tcc

```diff
@@ -0,0 +1,231 @@
+                        hashes[(idx-1)/2] = HashTreeT::get_hash(hash, std::next(it)->second);
+                        ++it;
+                    }
+                }
+            }
+
+            last_it = next_last_it;
+        }
+    }
+}
+
+template<typename FieldT, typename HashTreeT>
+FieldT merkle_tree_field<FieldT, HashTreeT>::get_value(const size_t address) const
+{
+    assert(libff::log2(address) <= depth);
+
+    auto it = values.find(address);
+    FieldT result = (it == values.end() ? FieldT("0") : it->second);
+
+    return result;
+}
+
+template<typename FieldT, typename HashTreeT>
+void merkle_tree_field<FieldT, HashTreeT>::set_value(const size_t address, const FieldT &value)
+{
+    assert(libff::log2(address) <= depth);
+    size_t idx = address + (1ul<<depth) - 1;
+
+    values[address] = value;
+    hashes[idx] = value;
+
+    for (int layer = depth-1; layer >=0; --layer)
+    {
+        idx = (idx-1)/2;
+
+        auto it = hashes.find(2*idx+1);
+        FieldT l = (it == hashes.end() ? hash_defaults[layer+1] : it->second);
+
+        it = hashes.find(2*idx+2);
+        FieldT r = (it == hashes.end() ? hash_defaults[layer+1] : it->second);
+
+        FieldT h = HashTreeT::get_hash(l, r);
+        hashes[idx] = h;
+    }
+}
+
+template<typename FieldT, typename HashTreeT>
+FieldT merkle_tree_field<FieldT, HashTreeT>::get_root() const
+{
+    auto it = hashes.find(0);
+    return (it == hashes.end() ? hash_defaults[0] : it->second);
+}
+
+template<typename FieldT, typename HashTreeT>
+merkle_authentication_path merkle_tree_field<FieldT, HashTreeT>::get_path(const size_t address) const
+{
+
+    // Create empty vector of size depth
+    merkle_authentication_path result(depth);
+
+    // Check that the node given has address within tree range
+    assert(libff::log2(address) <= depth);
+
+    //Compute node address on tree
+    size_t idx = address + (1ul<<depth) - 1;
+
+    // For each layer
+    for (size_t layer = depth; layer > 0; --layer)
+    {
+        //Compute the sibling node address and retrieve it
+        size_t sibling_idx = ((idx + 1) ^ 1) - 1;
+        auto it = hashes.find(sibling_idx);
+
+        // If last layer, retrieve sibling node value from vector values, else from vector hashes
+        if (layer == depth)
+        {
+            auto it2 = values.find(sibling_idx - ((1ul<<depth) - 1));
+            result[layer-1] = (it2 == values.end() ? FieldT("0") : it2->second);
+        }
+        else
+        {
+            result[layer-1] = (it == hashes.end() ? hash_defaults[layer] : it->second);
+        }
+
+
+        idx = (idx-1)/2;
+    }
+    std::reverse(result.begin(), result.end());
+    return result;
+}
+
+template<typename FieldT, typename HashTreeT>
+void merkle_tree_field<FieldT, HashTreeT>::dump() const
+{
+    std::cout << "hash default" << std::endl;
+    for (size_t i = 0; i <ZETH_MERKLE_TREE_DEPTH; i++)
```

Comment:
Why `ZETH_MERKLE_TREE_DEPTH` and not `depth`?

Reply:
I cannot see a reason. 

It's obviously a mistake made, riemann99 and I must have forgotten about `depth`.

Reply:
Fixed